### PR TITLE
feat(2fa): Add /recoveryCodes/exists endpoint

### PR DESCRIPTION
**Pull request by vpomerleau**
## Because

* We want to hook up the BackupCodeManager to the front end

## This pull request

* Add new /recoveryCodes/exists route
* Add new handler method that uses the BackupCodeManager to retrieve the count of remaining codes
* Add backup codes to account resolver
* Update fxa-settings account model to include backup codes count

## Issue that this pull request solves

Closes: #FXA-10231

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [ ] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

Any other information that is important to this pull request.

### CODE REVIEW

**Review by vbudhram — COMMENTED** [review]
@vpomerleau I think this might be the first time we have used libs in the auth-server that uses the database. Below is a patch that loads the settings page with the additional details

```
diff --git a/packages/fxa-auth-server/lib/routes/recovery-codes.js b/packages/fxa-auth-server/lib/routes/recovery-codes.js
index b06dd9d515..4ba74c5d5f 100644
--- a/packages/fxa-auth-server/lib/routes/recovery-codes.js
+++ b/packages/fxa-auth-server/lib/routes/recovery-codes.js
@@ -4,6 +4,8 @@
 
 'use strict';
 
+import { AppConfig } from '../types';
+
 const errors = require('../error');
 const isA = require('joi');
 const validators = require('./validators');
@@ -12,7 +14,8 @@ const RECOVERY_CODES_DOCS =
   require('../../docs/swagger/recovery-codes-api').default;
 const {
   BackupCodeManager,
-} = require('../../../../libs/accounts/two-factor/src/');
+} = require('@fxa/accounts/two-factor');
+import { setupAccountDatabase } from '@fxa/shared/db/mysql/account';
 
 const RECOVERY_CODE_SANE_MAX_LENGTH = 20;
 
@@ -20,7 +23,15 @@ module.exports = (log, db, config, customs, mailer, glean) => {
   const codeConfig = config.recoveryCodes;
   const RECOVERY_CODE_COUNT = (codeConfig && codeConfig.count) || 8;
 
-  const backupCodeManager = Container.get(BackupCodeManager);
+  let backupCodeManager;
+  (async () => {
+    const appConfig = Container.get(AppConfig);
+    const kyselyDb = await setupAccountDatabase(appConfig.database.mysql.auth);
+    if (!Container.has(BackupCodeManager)) {
+      Container.set(BackupCodeManager, new BackupCodeManager(kyselyDb));
+    }
+    backupCodeManager = Container.get(BackupCodeManager);
+  })();
 
   // Validate backup authentication codes
   const recoveryCodesSchema = validators.recoveryCodes(
@@ -146,11 +157,11 @@ module.exports = (log, db, config, customs, mailer, glean) => {
       async handler(request) {
         log.begin('checkRecoveryCodesExist', request);
 
-        const { sessionToken, uid } = request.auth.credentials;
+        const { email, uid } = request.auth.credentials;
 
         await customs.check(
           request,
-          sessionToken.email,
+          email,
           'checkRecoveryCodesExist'
         );
 
```

We could probably improve this code a bit.

File: packages/fxa-settings/src/models/Account.ts
Comment (vpomerleau) [reply packages/fxa-settings/src/models/Account.ts:195]: I wasn't certain here since in the libs and in the ux we call them backup codes...

File: packages/fxa-auth-server/lib/routes/backup-codes-manager.ts
Comment (vpomerleau) [inline packages/fxa-auth-server/lib/routes/backup-codes-manager.ts:10]: @vbudhram This is proving difficult to mock for tests, wondering if there might be a better way to set this up 🤔

File: packages/fxa-auth-server/bin/key_server.js
Comment (dschom) [inline packages/fxa-auth-server/bin/key_server.js:117]: Needs an await

Comment (vpomerleau) [reply packages/fxa-auth-server/bin/key_server.js]: 🙇‍♀️ thank you!!

File: packages/fxa-settings/src/components/Settings/UnitRowTwoStepAuth/index.tsx
Comment (vpomerleau) [inline packages/fxa-settings/src/components/Settings/UnitRowTwoStepAuth/index.tsx:57]: This console log is temporary and can be used to test that the number of remaining backup authentication codes is retrieved

File: packages/fxa-settings/src/models/Account.ts
Comment (vpomerleau) [reply packages/fxa-settings/src/models/Account.ts:195]: @vbudhram What are your thoughts on this - should I update all the things (from libs all the way to fxa-settings account model) to use the same naming convention? The naming difference between backend/frontend is tricky, in addition to the current planned user-facing name for the codes sent via SMS (recovery codes... 🥴 )

File: packages/fxa-auth-server/bin/key_server.js
Comment (dschom) [inline packages/fxa-auth-server/bin/key_server.js:49]: Should be `@fxa/shared/db/mysql/account`

File: packages/fxa-auth-server/bin/key_server.js
Comment (dschom) [inline packages/fxa-auth-server/bin/key_server.js:118]: nit: Maybe give this a better name? 

I also think something else might have landed on main that is already calling setupAccountDatabase. Double check before landing.

File: packages/fxa-auth-server/lib/routes/recovery-codes.js
Comment (dschom) [inline packages/fxa-auth-server/lib/routes/recovery-codes.js:148]: Does this action already exist?

File: packages/fxa-auth-server/test/client/api.js
Comment (dschom) [inline packages/fxa-auth-server/test/client/api.js:1176]: nit: New code we write can use async / await instead of promise chains.

File: packages/fxa-auth-server/test/local/routes/recovery-codes.js
Comment (dschom) [inline packages/fxa-auth-server/test/local/routes/recovery-codes.js:43]: tip: Using a sandbox can create a better isolation. See https://github.com/mozilla/fxa/pull/18038/files#r1855000025

File: packages/fxa-auth-server/test/remote/recovery_code_tests.js
Comment (dschom) [inline packages/fxa-auth-server/test/remote/recovery_code_tests.js:14]: Is it possible to undo the formatting, so we can see what changed? ie remove the indent?

**Comment by dschom** [comment]
@vpomerleau  Looking good. I'm ready to R+, please fix [this](https://github.com/mozilla/fxa/pull/17998#discussion_r1854998460), and looks like maybe you need a rebase on main?

File: packages/fxa-auth-server/lib/routes/recovery-codes.js
Comment (vpomerleau) [reply packages/fxa-auth-server/lib/routes/recovery-codes.js:148]: Added to `actions`, thank you!

**Comment by vpomerleau** [comment]
@dschom Thank you for the review! I updated with changes based on your comments. I initially started going down a rabbit hole of refactoring the recovery-codes route and test files but in the end opted to keep the changes minimal.

Wondering what your thoughts are on https://github.com/mozilla/fxa/pull/17998#discussion_r1850751082

File: packages/fxa-settings/src/models/Account.ts
Comment (dschom) [reply packages/fxa-settings/src/models/Account.ts:195]: I'd be inclined to to agree it should be `recoveryCodes`. The `auth-server` calls them recovery-codes and I feel like `auth-server` is actually the source of truth for this. However, it seems like the BackupCodeManager probably isn't named properly given it's predominantly used on the serverside? /not-sure 

It bother me less if the UI has different terms. Semantics of the UI and the underlying code will drift overtime. It's not ideal, but I don't know if it's worth changing. IMO, some comments can clear up any confusion in the front end. However it seems important to me that the semantics on the server side are consistent.

I wouldn't block on this though. It seems like a nice polish PR / followup to me. Just my two cents.

**Review by dschom — APPROVED** [review]

**Pull request merged**
